packages/at_client_mobile/lib/src/auth/at_auth_service_impl.dart

```diff
@@ -29,4 +127,436 @@ class AtAuthServiceImpl implements AtAuthService {
+        enrollmentId: atAuthKeys.enrollmentId);
+
+    await _keyChainManager.storeAtSign(atSign: atSignItem);
+  }
+
+  Future<void> _persistKeysLocalSecondary(AtAuthKeys? atAuthKeys) async {
+    if (atAuthKeys == null) {
+      throw AtException(
+          'Failed to store keys in Keychain manager for atSign: $_atSign. AtAuthKeys instance is null');
+    }
+
+    await _atClient!
+        .getLocalSecondary()!
+        .putValue(AtConstants.atPkamPublicKey, atAuthKeys.apkamPublicKey!);
+
+    // pkam private will not be available in case of secure element
+    if (atAuthKeys.apkamPrivateKey != null) {
+      await _atClient!
+          .getLocalSecondary()!
+          .putValue(AtConstants.atPkamPrivateKey, atAuthKeys.apkamPrivateKey!);
+    }
+
+    await _atClient!.getLocalSecondary()!.putValue(
+        AtConstants.atEncryptionPrivateKey,
+        atAuthKeys.defaultEncryptionPrivateKey!);
+
+    var updateBuilder = UpdateVerbBuilder()
+      ..atKey = AtKey.public('publickey', sharedBy: _atSign).build();
+    updateBuilder.atKey.metadata.ttr = -1;
+    updateBuilder.value = atAuthKeys.defaultEncryptionPublicKey;
+
+    await _atClient!
+        .getLocalSecondary()!
+        .executeVerb(updateBuilder, sync: true);
+
+    await _atClient!.getLocalSecondary()!.putValue(
+        AtConstants.atEncryptionSelfKey, atAuthKeys.defaultSelfEncryptionKey!);
+  }
+
+  Future<void> _init(AtChops atChops, {String? enrollmentId}) async {
+    await _initAtClient(atChops, enrollmentId: enrollmentId);
+    atLookUp!.atChops = atChops;
+    _atClient!.atChops = atChops;
+  }
+
+  Future<void> _initAtClient(AtChops atChops, {String? enrollmentId}) async {
+    AtClientManager atClientManager = AtClientManager.getInstance();
+    await atClientManager.setCurrentAtSign(
+        _atSign, _atClientPreference.namespace, _atClientPreference,
+        atChops: atChops,
+        serviceFactory: atServiceFactory,
+        enrollmentId: enrollmentId);
+    // ??= to support mocking
+    atLookUp ??= atClientManager.atClient.getRemoteSecondary()?.atLookUp;
+    atLookUp?.enrollmentId = enrollmentId;
+    atLookUp?.signingAlgoType = _atClientPreference.signingAlgoType;
+    atLookUp?.hashingAlgoType = _atClientPreference.hashingAlgoType;
+    _atClient ??= atClientManager.atClient;
+  }
+
+  @override
+  Future<AtEnrollmentResponse> enroll(EnrollmentRequest enrollmentRequest,
+      {String? keysFilePath}) async {
+    // Only one enrollment request can be submitted at a time.
+    // Subsequent requests cannot be submitted until the pending one is fulfilled.
+    var enrollmentStore = await _getEnrollmentStorage();
+    String? enrollmentInfoJsonString = await enrollmentStore.read();
+    // if enrollmentInfoJsonString is not null, it indicates that there is a pending
+    // enrollment request. So, do not allow another enrollment request.
+    if (enrollmentInfoJsonString != null) {
+      throw InvalidRequestException(
```

Comment:
can we through AtEnrollmentException here? InvalidRequestException is an exception from secondary server